awx/main/scheduler/task_manager_models.py

```diff
@@ -0,0 +1,121 @@
+# Copyright (c) 2022 Ansible by Red Hat
+# All Rights Reserved.
+import logging
+
+from django.conf import settings
+
+from awx.main.models import (
+    Instance,
+    InstanceGroup,
+)
+
+logger = logging.getLogger('awx.main.scheduler')
+
+
+class TaskManagerInstance:
+    """A class representing minimal data the task manager needs to represent an Instance."""
+
+    def __init__(self, obj=None, node_type=None, capacity=None, hostname=None):
+        self.obj = obj
+        self.node_type = node_type
+        self.remaining_capacity = capacity
+        self.capacity = capacity
+        self.hostname = hostname
+
+
+class TaskManagerInstances:
+    def __init__(self, active_tasks, instances=None):
+        self.instances_by_hostname = dict()
+        self._remaining_capacity_updated = False
+        if instances is None:
+            instances = (
+                Instance.objects.filter(hostname__isnull=False, enabled=True).exclude(node_type='hop').only('node_type', 'capacity', 'hostname', 'enabled')
+            )
+        instance_list = [
+            TaskManagerInstance(
+                obj=instance,
+                node_type=instance.node_type,
+                capacity=instance.capacity,
+                hostname=instance.hostname,
+            )
+            for instance in instances
+        ]
+        self._update_remaining_capacity(active_tasks)
+        for instance in instance_list:
+            self.instances_by_hostname[instance.hostname] = instance
+
+    def _update_remaining_capacity(self, tasks):
+        """Takes a list of tasks and updates the remaining capacity on our TaskManagerInstance list.
+
+        Computes remaining capacity for all the instances based on currently running and waiting tasks.
+        """
+        if self._remaining_capacity_updated:
+            # Only want to do this once
+            return
+        for task in tasks:
+            if task.status not in ['waiting', 'running']:
+                continue
+            control_instance = self.instances_by_hostname.get(task.controller_node, '')
+            execution_instance = self.instances_by_hostname.get(task.execution_node, '')
+            if execution_instance and execution_instance.node_type in ('hybrid', 'execution'):
+                self.instances_by_hostname[task.execution_node].remaining_capacity -= task.task_impact
+            if control_instance and control_instance.node_type in ('hybrid', 'control'):
+                self.instances_by_hostname[task.controller_node].remaining_capacity -= settings.AWX_CONTROL_NODE_TASK_IMPACT
+
+    def __getitem__(self, hostname):
+        return self.instances_by_hostname.get(hostname)
+
+    def __contains__(self, hostname):
+        return hostname in self.instances_by_hostname
+
+
+class TaskManagerInstanceGroups:
+    """A class representing minimal data the task manager needs to represent an InstanceGroup."""
+
+    def __init__(self, instances_by_hostname=None, instance_groups=None):
```

Comment:
super minor nit about the kwargs, if someone uses the defaults like `TaskManagerInstanceGroups()` then I think that would give an error due to iterating over the `None` value. But in this case, making it required may be bad, because someone could give `instance_groups` by itself.